core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/GroupUniqueIDDomainResolver.java

```diff
@@ -91,26 +100,37 @@ public String getDomainForGroupId(String groupId, int tenantId) throws UserStore
 
             // Read the cache first.
             String domainName = uniqueIdDomainCache.get(groupId);
-            if (domainName != null) {
+
+            if (org.apache.commons.lang.StringUtils.isBlank(domainName)) {
+                // Domain name is not in the cache.
                 if (log.isDebugEnabled()) {
-                    log.debug("Cache hit for group id: " + groupId);
+                    log.debug("Cache miss for group id: " + groupId + " searching from the database");
+                }
+                // Read the domain name from the Database;
+                domainName = getDomainFromDB(groupId, tenantId);
+                // Update the cache.
+                if (org.apache.commons.lang.StringUtils.isNotBlank(domainName)) {
+                    uniqueIdDomainCache.put(groupId, domainName);
+                    if (log.isDebugEnabled()) {
+                        log.debug("Domain with name: " + domainName + " retrieved from the database.");
+                    }
                 }
-                return domainName;
-            }
-            // Domain name is not in the cache.
-            if (log.isDebugEnabled()) {
-                log.debug("Cache miss for group id: " + groupId + " searching from the database");
             }
-            // Read the domain name from the Database;
-            domainName = getDomainFromDB(groupId, tenantId);
-            // Update the cache.
-            if (domainName != null) {
-                uniqueIdDomainCache.put(groupId, domainName);
-                if (log.isDebugEnabled()) {
-                    log.debug("Domain with name: " + domainName + " retrieved from the database.");
+            if (org.apache.commons.lang.StringUtils.isNotBlank(domainName) &&
+                    !UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME.equals(domainName)) {
+                RealmService realmService = UserCoreUtil.getRealmService();
+                UserStoreManager userStoreManager =
+                        ((AbstractUserStoreManager) realmService.getTenantUserRealm(tenantId)
+                                .getUserStoreManager()).getSecondaryUserStoreManager(domainName);
+                if (userStoreManager == null) {
+                    deleteDomainFromDB(domainName, groupId, tenantId);
```

Comment:
let's add debug logs, stating that the entries are outdated hence removing them from the cache and DB.

Reply:
Fixed: 6c08cef558b386ba44af7d826510ac5d53840a43